### ISSUE TITLE
sqrt and --no-ieee-float mystery on SLES 12

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#24112]
PRs #24106 and #24111 add a .suppressif for `check-sqrtr-noieee`.

After PR #24066, the new test `check-sqrt` was failing on some systems, but not others, when using `--no-ieee-float`. `--no-ieee-float` translates to `-ffast-math` and should enable more optimization. However, on the affected systems, compiling with `--no-ieee-float --fast` fails to inline the `sqrt` call into an instruction on that system.

<details>

```
$ chpl check-sqrt-noieee.chpl  --llvm-print-ir multiSqrt --llvm-print-ir-stage asm --fast --no-ieee-float

1 symbol names to disassemble


# Disassembling symbol multiSqrt_chpl


/tmp/chpl-mferguson.deleteme-K1RtoM/chpl__module.o:	file format elf64-x86-64

Disassembly of section .text:

00000000000301d0 <multiSqrt_chpl>:
   301d0: 50                           	pushq	%rax
   301d1: c5 f9 2e 05 00 00 00 00      	vucomisd	, %xmm0 <multiSqrt_chpl+0x9>
   301d9: 76 14                        	jbe	 <L0>
   301db: 0f 1f 44 00 00               	nopl	(%rax,%rax)
<L2>:
   301e0: e8 00 00 00 00               	callq	 <L1>
<L1>:
   301e5: c5 f9 2e 05 00 00 00 00      	vucomisd	, %xmm0 <multiSqrt_chpl+0x1d>
   301ed: 77 f1                        	ja	 <L2>
<L0>:
   301ef: 58                           	popq	%rax
   301f0: c3                           	retq
   301f1: 66 66 66 66 66 66 2e 0f 1f 84 00 00 00 00 00 	nopw	%cs:(%rax,%rax)
```

(note that the `callq` above is calling `sqrt`; we just don't see `sqrt` mentioned due to not using `llvm-objdump -r` due to https://github.com/llvm/llvm-project/issues/75539 )

</details>

At the same time, on the affected systems, compiling a similar C program with the `clang` we are configured to use shows the expected inlining:

<details>

``` c
// test.c
#include <math.h>
#include <stdio.h>

__attribute__((noinline))
static double multiSqrt(double n) {
  while (n > 10.0) {
    n = sqrt(n);
  }
  return n;
}

int main(int argc, char** argv) {
  double nums[1000];
  double sum = 0.0;
  for (int i = 0; i < 1000; i++) {
    nums[i] = multiSqrt((double)i);
  }

  for (int i = 0; i < 1000; i++) {
    sum += nums[i];
  }

  printf("sum is %f\n", sum);
  return 0;
}
```

```
$ /data/cf/chapel/chpl-deps/spack/opt/spack/linux-sles12-haswell/gcc-12.2.0/llvm-15.0.4-k67u3v54wddzwta5ecwljfvjh4lxkrnm/bin/clang test.c -O3 -ffast-math -fno-math-errno -S
$ cat test.s
	.text
	.file	"test.c"
	.globl	main                            # -- Begin function main
	.p2align	4, 0x90
	.type	main,@function

...

	.section	.rodata.cst8,"aM",@progbits,8
	.p2align	3                               # -- Begin function multiSqrt
.LCPI1_0:
	.quad	0x4024000000000000              # double 10
	.text
	.p2align	4, 0x90
	.type	multiSqrt,@function
multiSqrt:                              # @multiSqrt
	.cfi_startproc
# %bb.0:
	ucomisd	.LCPI1_0(%rip), %xmm0
	jbe	.LBB1_3
# %bb.1:
	movsd	.LCPI1_0(%rip), %xmm1           # xmm1 = mem[0],zero
	.p2align	4, 0x90
.LBB1_2:                                # =>This Inner Loop Header: Depth=1
	sqrtsd	%xmm0, %xmm0
	ucomisd	%xmm1, %xmm0
	ja	.LBB1_2
.LBB1_3:
	retq
.Lfunc_end1:
	.size	multiSqrt, .Lfunc_end1-multiSqrt
	.cfi_endproc
                                        # -- End function
...
```

Here we can see that there is no `callq` in the body of `multiSqrt`, as expected.

</details>

It is a mystery to me why this is occurring. The problem seems specific to OS version rather than processor type. In particular, all of the affected systems are running SUSE Linux Enterprise Server 12 SP3, which comes with GCC 4.8.5. It does not seem to be an issue on SUSE Linux Enterprise Server 15 SP3, which comes with GCC 7.5.0. Similarly, it does not seem to be an issue on Ubuntu 23.10 which comes with GCC 13.2.0. It is also possible that it has to do with how the system LLVM and clang were built on the affected systems.

**Comment by milthorpe** [comment]
I have also observed this problem when compiling [miniBUDE](https://github.com/milthorpe/miniBUDE/) with Chapel 1.33 (and pre-release Chapel 2.0) on three different systems: NCI Gadi (Rocky Linux 8.9), ExCL leconte (CentOS 8.3) and ExCL thunderx (CentOS Stream 8). The specific line that's causing the problem is a [call to `sqrt` inside an otherwise-vectorizable `foreach` loop](https://github.com/milthorpe/miniBUDE/blob/9f3fec5594bdd5c7830fca19571139be3682e87d/src/chapel/Bude.chpl#L635); if this call is present, I get the following warning compiling with `chpl --fast --no-ieee-float --ccflags -fno-math-errno --llvm-remarks=vector`:

```
Bude.chpl:516:0: opt analysis for 'loop-vectorize' - loop not vectorized: library call cannot be vectorized. Try compiling with -fno-math-errno, -ffast-math, or similar flags
```

On all Ubuntu systems I've tried (Ubuntu 22.04 on multiple Intel and AMD generations, Ubuntu 20.04 on Intel Skylake), this loop vectorizes correctly without any warnings.

A possible clue: the problematic systems are all using an older version of glibc (2.28). 

```
[gadi-login-09 ~]$ /lib64/libc.so.6
GNU C Library (GNU libc) stable release version 2.28.
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 8.5.0 20210514 (Red Hat 8.5.0-20).
libc ABIs: UNIQUE IFUNC ABSOLUTE
For bug reporting instructions, please see:
<http://www.gnu.org/software/libc/bugs.html>.
```

Ubuntu 22.04 is using glibc 2.35; Ubuntu 20.04 is using glibc 2.31. I note that there were [changes in the definitions of <math.h> functions in glibc 2.31](https://sourceware.org/legacy-ml/libc-announce/2020/msg00001.html), but I haven't looked into the details.

**Comment by mppf** [comment]
@milthorpe -- thanks for your note & I think that the glibc version is a good guess as to the nature of the problem.

 * SUSE Linux Enterprise Server 12 SP3 has glibc 2.22
 * SUSE Linux Enterprise Server 15 SP3 has glibc 2.31

As with your information: it seems like something in glibc changed in a relevant way between 2.28 and 2.31.

Do you think we should be looking for a workaround to this problem? Do you already know of one? (I made this issue about a test failure that is pretty low-priority, but the fact that you ran into it makes me wonder if we should give it more attention).

Perhaps we could suggest using a Spack install of Chapel (which I think, but am not certain, will build a new glibc)? Or maybe suggest using the C backend on such systems?

**Comment by milthorpe** [comment]
I'm actively looking for a workaround for this issue as the lack of vectorization causes a roughly 2-8x slowdown for miniBUDE. I have tried and failed to install a newer glibc from Spack on all three systems.

**Comment by jabraham17** [comment]
I suspect this can be solved (at least with the LLVM backend) by explicitly calling the LLVM sqrt intrinsic, rather than relying on a glibc library call being optimized.

[This branch](https://github.com/chapel-lang/chapel/compare/main...jabraham17:chapel:sqrt-llvm) takes a first pass at this by just sticking in a new primitive "mysqrt", adapted from the "fma" primitive. If people like this approach I can clean this up and open a PR.

I tested this on a SUSE 12 SP3 system with glibc 2.22 both by running `start_test test/llvm/no-math-errno/check-sqrt-noieee.chpl` (it passed) and manually examining the generated assembly (inlined sqrt inst)

**Comment by milthorpe** [comment]
@jabraham17 Something like this will work well - thanks! I tested miniBUDE against a modified version of your changes so that both the `sqrt` and `abs` functions are implemented as vector instructions, and it improved the performance by more than 8x. (The loop is now properly vectorized.)
I note that there is already a `PRIM_SQRT` in the compiler code, which may have been intended for this purpose, but it doesn't appear to be 'plumbed in'.

**Comment by mppf** [comment]
@milthorpe @jabraham17 that's great that we've got some sort of solution here, thanks to both of you!

> I note that there is already a `PRIM_SQRT` in the compiler code, which may have been intended for this purpose, but it doesn't appear to be 'plumbed in'.

It was added for param (compile-time) sqrt calls but IMO it's reasonable to also arrange for it to call the LLVM intrinsic

Also, re. emitSqrtForC & similar fabs function emitting an error for the C backend -- I think we'll probably need this to call the relevant C function.
